Diagnostic to flag cases where control flow can break content projection

crisbeto opened the pull request:
These changes are a follow-up to the fix from #52414. They add a new diagnostic to the compiler that will flag cases where an element would've been projected into a specific slot of a component, but it isn't because it exists inside of a control flow node that has multiple root nodes. Aside from the diagnostic, I have to move some code around so it can be reused. Includes the following commits:

### refactor(compiler): expose utility for creating CSS selectors from AST nodes
When doing directive matching in the compiler, we need to be able to create a selector from an AST node. We already have the utility, but these changes simplify the public API and expose it so it can be used in `compiler-cli`.

### refactor(compiler-cli): expose ng-content selectors and preserveWhitespaces during template type checking
These changes expose the `ngContentSelectors` and `preserveWhitespaces` metadata to the TCB so they can be used in the next commit to implement a new diagnostic.

### test(core): add tests for control flow content projection with ng-container
The control flow projection diagnostic will mention `ng-container` as a workaround for projection multiple nodes. These changes add a couple of tests to ensure that the approach works.

### fix(compiler-cli): add diagnostic for control flow that prevents content projection
This is a follow-up to the fix from https://github.com/angular/angular/pull/52414. It adds a diagnostic that will tell users when a control flow is preventing its direct descendants from being projected into a specific component slot.

crisbeto committed:
refactor(compiler): expose utility for creating CSS selectors from AST nodes

When doing directive matching in the compiler, we need to be able to create a selector from an AST node. We already have the utility, but these changes simplify the public API and expose it so it can be used in `compiler-cli`.

crisbeto committed:
refactor(compiler-cli): expose ng-content selectors and preserveWhitespaces during template type checking

These changes expose the `ngContentSelectors` and `preserveWhitespaces` metadata to the TCB so they can be used in the next commit to implement a new diagnostic.

crisbeto committed:
test(core): add tests for control flow content projection with ng-container

The control flow projection diagnostic will mention `ng-container` as a workaround for projection multiple nodes. These changes add a couple of tests to ensure that the approach works.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -905,6 +905,105 @@ class TcbDomSchemaCheckerOp extends TcbOp {
+ * of a control flow expression *would have been projected* into a specific slot, if the control
+ * flow node didn't exist.
+ */
+class TcbControlFlowContentProjectionOp extends TcbOp {
```
I wasn't totally sure if another `TcbOp` was the right place to put this since it doesn't add any statements to the TCB, but I saw that we have some precedent for cases in the `TcbDomSchemaCheckerOp`. My main concern was to isolate the check from everything else in the TCB. The other option was to put it in a private method inside the `Scope`, but that class is pretty large as it is.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -340,6 +348,34 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+          blockName} has more than one node at its root. To project the node in the right slot, you can:\n`,
+      `1. Wrap the content of the ${blockName} block in an <ng-container/> that matches the "${
+          slotSelector}" selector.`,
+      `2. Split the content of the ${blockName} block into across multiple ${
```
`into across` -> either of those words, not both (I don't know if there's a preference towards one of those)

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -340,6 +348,34 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+    if (preservesWhitespaces) {
+      lines.push(
+          `Note: the host component has \`preserveWhitespaces: true\` which may ` +
+          `cause whitespace to affect content projection.`);
```
WDYT about making this more targeted towards the situation where we know it's affected, i.e. detect when it is actually just whitespace that is affecting the result and then reporting smth like:

> Note: the host component has `preserveWhitespaces: true` and ${projectionNode.tagName} has surrounding whitespace, which affects content projection.

crisbeto replied:
I was trying to avoid putting in too many special cases for `preserveWhitespaces: true` since it isn't commonly used.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -340,6 +348,34 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+
+    this._diagnostics.push(makeTemplateDiagnostic(
+        templateId, this.resolver.getSourceMapping(templateId), projectionNode.startSourceSpan,
+        ts.DiagnosticCategory.Error,
```
Should this really be an error? I'd have expected a warning, possibly even an extended diagnostic.

I also think the messaging needs to include a way to suppress this diagnostic if the template is intentional.

crisbeto replied:
I think the way to suppress it would be to remove whatever makes the element project into the slot. As for having it be an error versus a warning, I won't feel strongly about it.

crisbeto replied:
Switched to a warning based on the discussion.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -905,6 +905,105 @@ class TcbDomSchemaCheckerOp extends TcbOp {
+  }
+
+  private findPotentialControlFlowNodes() {
+    const result: (TmplAstIfBlockBranch|TmplAstForLoopBlock)[] = [];
```
The [TS style guide](https://google.github.io/styleguide/tsguide.html#arrayt-type) prescribes the usage of the `Array` type for complex generic types instead of the shorthand `[]` syntax.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -905,6 +905,105 @@ class TcbDomSchemaCheckerOp extends TcbOp {
+      // Only `@for` blocks and the first branch of `@if` blocks participate in content projection.
+      // Skip nodes with less than two children since it's impossible for them to run into the
+      // issue that we're checking for.
+      if (child instanceof TmplAstForLoopBlock && child.children.length > 1) {
+        eligibleNode = child;
+      } else if (child instanceof TmplAstIfBlock && child.branches[0].children.length > 1) {
+        eligibleNode = child.branches[0];
+      }
+
+      if (eligibleNode === null) {
+        continue;
+      }
```
nit: I'd remove the `children.length > 1` from those conditions and move it to a `eligibleNode === null || eligibleNodes.children.length <= 1` condition

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -905,6 +905,105 @@ class TcbDomSchemaCheckerOp extends TcbOp {
+          // content projection behavior.
+          if (this.tcb.hostPreserveWhitespaces) {
+            count++;
+          } else {
+            count += node.value.trim().length > 0 ? 1 : 0;
+          }
```
nit:

```
} else if (node.value.trim().length > 0) {
  count++;
}
```

at which point it could even become a single condition. Happy with the current code as well though, just an observation.

JoostK replied:
I do wonder how this works for e.g.:

```html
<span>@if (true) { <my-cmp/> }</span>
```

Even with `preserveWhitespaces: false` I'd expect the whitespace to be retained, since removing it would remove a space from the HTML which could be load-bearing (and IFAIK `preserveWhitespaces` is conservative in those cases, retaining whitespace that may be load-bearing.)

crisbeto replied:
I haven't checked it, but my understanding is that the example you posted won't retain the whitespace, unless `preserveWhitespaces` is enabled. However, if it was a data binding instead of `<my-cmp/>`, then it will retain the whitespace (instruction becomes something like `interpolation(" " + expr + " ")`.

JoostK replied:
Ah that could be the case, in that case we're okay since those can't be projected anyway.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -340,6 +348,34 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+    const blockName = controlFlowNode instanceof TmplAstIfBlockBranch ? '@if' : '@for';
+    const lines = [
+      `Node matches the "${slotSelector}" slot of the "${
+          componentName}" component, but will not be projected because the surrounding ${
```
"will not be projected" may be inaccurate if there is a default slot. Perhaps this should have two forms ("will not be projected" and "will be projected into the default projection slot" or smth like that) or generalize to e.g. "will not be projected there".

crisbeto replied:
I changed the wording a bit to mention that it won't be projected into that slot.

crisbeto commented:
I've addressed most of the feedback and will bring up the discussion around using a warning versus an error later today.

jessicajaniuk reviewed (approved):
LGTM. I'm indifferent on warning vs error.

reviewed-for: fw-core, public-api

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/oob.ts:
```diff
@@ -340,6 +348,34 @@ export class OutOfBandDiagnosticRecorderImpl implements OutOfBandDiagnosticRecor
+    const blockName = controlFlowNode instanceof TmplAstIfBlockBranch ? '@if' : '@for';
+    const lines = [
+      `Node matches the "${slotSelector}" slot of the "${
+          componentName}" component, but will not be projected into the specifc slot because the surrounding ${
```
```suggestion
          componentName}" component, but will not be projected into the specific slot because the surrounding ${
```

I approved but missed this update, small typo here

crisbeto committed:
fix(compiler-cli): add diagnostic for control flow that prevents content projection

This is a follow-up to the fix from #52414. It adds a diagnostic that will tell users when a control flow is preventing its direct descendants from being projected into a specific component slot.

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-core

AndrewKushnir commented:
This PR was merged into the repository by commit b4d022e230ca141b12437949d11dc384bfe5c082.
